### DISCUSSION TITLE
:q[uit] to drop changes

### DISCUSSION THREAD

**Discussion by franklouwers** [LazyVim/LazyVim#1399] [Q&A] [answered]
Hi,

I used to be a one-buffer-only vimmer. I am starting to get the hang of multiple buffers, but there's one thing I find extremely annoying and it involves `:q`. I hope there's an easy solution for this.

Imaging you have a dir with multiple files, including "lib.rb", "helper.rb".

1. You open the Neo-Browser, and decide you want to edit `helper.rb`.
2. You start hacking, then realise this is the wrong place, 
3. use `:q` to close the ~file~ buffer, not saving the file, 
4. and change something in `lib.rb`.
5. One hour later, when working on something else, you open `helper.rb` again, and add a routine. 
6. You save `helper.rb`.

What just happend? Well, your old code is still in there, the code you didn't intend to save. (the one from step 2 above).

I understand this is because you didn't actually close-and-discard-changes in step 3, but it annoys the hell out of me. How can I make `:quit` work like I am used to, as in "Ask me to save change, and if I enter `n`, drop the changes for good.

**Comment by jyuan0** [comment]
By default, `:q` on a buffer with unsaved changes should ask you whether you want to save the changes (the behavior you're used to). Maybe you have an auto-save plugin installed?

**Reply by franklouwers** [reply]
It does only when neo-tree isn't open. When it's open, it "closes" the buffer (basically hides it) and when I re-open the file from neo-tree, it just re-opens the existing buffer with that file (including the unsaved changes).

**Reply by dpetka2001** [reply]
In my experience, this happens because `:q` closes a window and doesn't unload the buffer. In order to unload the buffer and thus getting the prompt to save the changes, you would have to `:bdelete` which deletes the buffer and unloads it. `:q` only prompts for saving or not the changes only if the window is the last window. If you follow the steps of your workflow, after closing the window with `:q`, you should still see in the neo-tree window that is still open (since you said this only happens when you have neo-tree open) an icon that suggests that file has unsaved changes.

**Reply by franklouwers** [reply]
Exactly. So is there a way to change the behaviour of `:q` to perform a `:bDelete` (it needs the `force` option as there are unsaved changes in the buffer) as well?

**Reply by dpetka2001** [reply]
You should just delete the buffer. `:q` is for interacting with windows, not buffers. If it's the last window (which prompts for saving changes or not) is just a byproduct of closing Neovim. `:h :q` also mentions that a `QuitPre` event is triggered, so you might want to add an autocommand on that event that does what you want to do.

**Reply by dpetka2001** [reply]
Or alternatively you could use `:q!` which doesn't write any changes and seems to behave like you want.

**Reply by franklouwers** [reply]
It's weird that `:q` and `:q!` behave so differently. I guess I want `:q` to behave exactly like `:q!`, but with asking me to save before closing (or as a second option, complaining when there are unsaved changes)

**Reply by dpetka2001** [reply, marked as answer]
That's what `:bdelete` is for. It's for handling buffers and their context. `:q`is for handling windows. Those are seperate terms. I tried with an autocommand and it seems to do what you want. Try the following in `/lua/config/autocmds.lua`
```lua
vim.api.nvim_create_autocmd("QuitPre", {
  pattern = "*",
  command = "bdelete",
  group = vim.api.nvim_create_augroup("DeleteBufferOnQuit", { clear = true }),
  desc = "Delete buffer when quitting",
})
```
Test it out to see if it fits your needs.

**Reply by franklouwers** [reply]
Thanks!
